docs/combinatorics.rst

```diff
@@ -1,6 +1,210 @@
+:meth:`Tree.rank` method.
+
+.. code-block:: python
+
+    ranks = [t.rank() for t in tskit.all_trees(num_leaves=3)]
+    print("Ranks of 3-leaf trees:", ranks)
+
+::
+
+   Ranks of 3-leaf trees: [(0, 0), (1, 0), (1, 1), (1, 2)]
+
+.. note::
+    Ranks in combinatorics are typically natural numbers. However,
+    we refer to this tuple of shape and label rank as a rank because
+    it serves the same purpose of indexing trees in an enumeration.
+
+For details on how shapes and labellings are ordered, see
+:ref:`sec_enumerating_topologies`.
+
+We can also reconstruct a leaf-labelled tree given its rank. This process
+is known as unranking, and can be performed using the :meth:`Tree.unrank`
+method.
+
+.. code-block:: python
+
+    for rank in [(0, 0), (1, 0), (1, 1), (1, 2)]:
+        t = Tree.unrank(rank, num_leaves=3)
+        display(SVG(t.draw(node_labels={0: 0, 1: 1, 2: 2}, order="tree")))
+
+.. image:: _static/topology_0_0.svg
+    :width: 24%
+.. image:: _static/topology_1_0.svg
+    :width: 24%
+.. image:: _static/topology_1_1.svg
+    :width: 24%
+.. image:: _static/topology_1_2.svg
+    :width: 24%
+
+++++++++
+Examples
+++++++++
+
+One application of tree ranks is to count the different
+leaf-labelled topologies in a tree sequence. Since the ranks
+are just tuples, we can use a Python ``Counter`` to track them.
+Here, we count and unrank the most frequently seen
+topology in a tree sequence. For brevity, this example assumes
+samples are synonymous with leaves.
+
+.. code-block:: python
+
+    rank_counts = collections.Counter(t.rank() for t in ts.trees())
+    most_freq_rank, count = rank_counts.most_common(1)[0]
+    Tree.unrank(most_freq_rank, num_leaves=ts.num_samples())
+
+.. _sec_enumerating_topologies:
+
+++++++++++++++++++++++
+Enumerating Topologies
+++++++++++++++++++++++
+
+This section expands briefly on the approach used to enumerate
+tree topologies that serves as the basis for :meth:`Tree.rank`
+and :meth:`Tree.unrank`.
+To enumerate all rooted, leaf-labelled tree topologies, we first
+formulate a system of ordering and enumerating tree shapes. Then
+we define an enumeration of labellings given an arbitrary tree shape.
+
+***********************
+Enumerating Tree Shapes
+***********************
+
+Starting with :math:`n = 1`, we see that the only shape for a tree
+with a single leaf is a single root leaf. A tree with :math:`n > 1`
+leaves can be obtained by joining at least two trees whose number of
+leaves sum to :math:`n`.
+This maps very closely to the concept of integer partitions.
+Each tree shape of :math:`n` leaves can be represented by taking a
+nondecreasing integer partition of :math:`n` (elements of the partition
+are sorted in nondecreasing order) and recursively partitioning its
+elements. The order in which we select partitions of :math:`n` is
+determined by the efficient
+`rule_asc <http://jeromekelleher.net/generating-integer-partitions.html>`_
+algorithm for generating them.
```

Comment:
Why not insert your nice explanation from before of the order that the integer partitions come in?  (I think you had such a description?)